authexternalbrowser.go

```diff
@@ -189,67 +189,64 @@ func authenticateByExternalBrowser(
 		return nil, nil, err
 	}
 
-	var encodedSamlResponse string
-	var acceptErr error
-	var tokenErr error
-	acceptErr = nil
-	resultChan := make(chan string)
+	encodedSamlResponseChan := make(chan string)
+	errChan := make(chan error)
 
-	go func(l net.Listener) {
-		for {
-			conn, err := l.Accept()
-			if err != nil {
-				glog.V(1).Infof("unable to accept connection. err: %v", err)
-				log.Fatal(err)
-				return
-			}
-			go func(c net.Conn) {
-				var buf bytes.Buffer
-				total := 0
-				for {
-					b := make([]byte, bufSize)
-					n, err := c.Read(b)
-					if err != nil {
-						if err != io.EOF {
-							glog.V(1).Infof("error reading from socket. err: %v", err)
-							acceptErr = err
+	var encodedSamlResponse string
+	var errFromGoroutine error
+	for {
+		conn, err := l.Accept()
+		if err != nil {
+			glog.V(1).Infof("unable to accept connection. err: %v", err)
+			log.Fatal(err)
+		}
+		fmt.Println("Hello")
+		go func(c net.Conn) {
+			var buf bytes.Buffer
+			total := 0
+			encodedSamlResponse := ""
+			var errAccept error
+			for {
+				b := make([]byte, bufSize)
+				n, err := c.Read(b)
+				if err != nil {
+					if err != io.EOF {
+						glog.V(1).Infof("error reading from socket. err: %v", err)
+						errAccept = &SnowflakeError{
+							Number:      ErrFailedToGetExternalBrowserResponse,
+							SQLState:    SQLStateConnectionRejected,
+							Message:     errMsgFailedToGetExternalBrowserResponse,
+							MessageArgs: []interface{}{err},
 						}
-						break
-					}
-					total += n
-					buf.Write(b)
-					if n < bufSize {
-						// We successfully read all data
-						s := string(buf.Bytes()[:total])
-						encodedSamlResponse, tokenErr = getTokenFromResponse(s)
-						break
 					}
-					buf.Grow(bufSize)
+					break
 				}
-				if encodedSamlResponse != "" {
-					httpResponse := buildResponse(application)
-					c.Write(httpResponse.Bytes())
-					resultChan <- encodedSamlResponse
+				total += n
+				buf.Write(b)
+				if n < bufSize {
+					// We successfully read all data
+					s := string(buf.Bytes()[:total])
+					encodedSamlResponse, errAccept = getTokenFromResponse(s)
+					break
 				}
-				c.Close()
-			}(conn)
-		}
-	}(l)
-
-	encodedSamlResponse = <-resultChan
-	l.Close()
+				buf.Grow(bufSize)
+			}
+			if encodedSamlResponse != "" {
+				httpResponse := buildResponse(application)
+				c.Write(httpResponse.Bytes())
+			}
+			c.Close()
+			encodedSamlResponseChan <- encodedSamlResponse
+			errChan <- errAccept
+		}(conn)
 
-	if tokenErr != nil {
-		return nil, nil, tokenErr
+		encodedSamlResponse = <-encodedSamlResponseChan
+		errFromGoroutine = <-errChan
+		break
```

Comment:
`authexternalbrowser.go:245:3: the surrounding loop is unconditionally terminated (SA4004)`

Reply:
ah, fixed in the following changes.